helix-core/src/indent.rs

```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
-/// This is because the indentation is only increased starting at the second line of the node.
-fn calculate_indentation(
-    query: &IndentQuery,
-    node: Option<Node>,
-    line: usize,
-    newline: bool,
-) -> usize {
-    let mut increment: isize = 0;
+// This assumes that the kind matches and checks for all the other conditions
+fn matches<'a>(query_node: &IndentQueryNode, node: Node<'a>, cursor: &mut TreeCursor<'a>) -> bool {
+    match query_node {
+        IndentQueryNode::SimpleNode(_) => true,
+        IndentQueryNode::ComplexNode {
+            kind: _,
+            kind_not_in,
+            parent_kind_in,
+            field_name_in,
+        } => {
+            if !kind_not_in.is_empty() {
+                let kind = node.kind();
+                if kind_not_in.iter().any(|k| k == kind) {
+                    return false;
+                }
+            }
+            if !parent_kind_in.is_empty() {
+                let parent_matches = node.parent().map_or(false, |p| {
+                    let parent_kind = p.kind();
+                    parent_kind_in
+                        .iter()
+                        .any(|kind| kind.as_str() == parent_kind)
+                });
+                if !parent_matches {
+                    return false;
+                }
+            }
+            if !field_name_in.is_empty() {
+                let parent = match node.parent() {
+                    None => {
+                        return false;
+                    }
+                    Some(p) => p,
+                };
+                cursor.reset(parent);
+                debug_assert!(cursor.goto_first_child());
+                loop {
+                    if cursor.node() == node {
+                        if let Some(cursor_name) = cursor.field_name() {
+                            if !field_name_in.iter().any(|n| n == cursor_name) {
+                                return false;
+                            }
+                        } else {
+                            return false;
+                        }
+                        break;
+                    }
+                    debug_assert!(cursor.goto_next_sibling());
+                }
+            }
+            true
+        }
+    }
+}
 
-    let mut node = match node {
-        Some(node) => node,
-        None => return 0,
-    };
+fn contains_match<'a>(
+    scope: &[IndentQueryNode],
+    node: Node<'a>,
+    cursor: &mut TreeCursor<'a>,
+) -> bool {
+    let current_kind = node.kind();
+    let first = scope.partition_point(|n| n.kind() < Some(current_kind));
+    for named_node in scope[first..]
+        .iter()
+        .take_while(|n| n.kind() == Some(current_kind))
+    {
+        if matches(named_node, node, cursor) {
+            return true;
+        }
+    }
+    for unnamed_node in scope.iter().take_while(|n| n.kind().is_none()) {
+        if matches(unnamed_node, node, cursor) {
+            return true;
+        }
+    }
+    false
```

Comment:
This can be simplified with `any`

Reply:
Yeah, `any` is nice. I updated the part with the named nodes above as well to use it.